ARROW-4386: [Rust] Temporal array support

nevi-me opened the pull request:
This adds support for date and time arrays

nevi-me commented:
@andygrove @paddyhoran @sunchao please have a look when you get time. I'll leave inline comments for questions of my own and places where I need a helping hand.

nevi-me commented on rust/arrow/src/datatypes.rs:
```diff
@@ -156,6 +157,107 @@ make_type!(UInt64Type, u64, DataType::UInt64, 64, 0u64);
+}
+
+make_temporal_type!(
+    TimestampSecondType,
```
Turns out I can't create a single `TimestampType` that can cater for all `TimeUnit` enum variants due to:

```rust
fn get_data_type() -> DataType {
    $data_ty
}
```

This is because when implementing the above trait member, I can't know at runtime what `TimeUnit` the array is implementing.
Is there a better/cleaner way of doing this?

sunchao replied:
I'm thinking that similar to the C++ implementation, perhaps it's better to have the `TimeUnit` as a field to the `TimestampType`. For this, we'll need to change `DataType` so that date, time, and timestamp types won't carry unit.

Perhaps we can address this in a follow-up PR though.

nevi-me commented on rust/arrow/src/datatypes.rs:
```diff
@@ -156,6 +157,107 @@ make_type!(UInt64Type, u64, DataType::UInt64, 64, 0u64);
+    0i64
+);
+make_temporal_type!(
+    TimestampMillisecondType,
```
These names ended up becoming long. Would we prefer `TimestampType + TimestampMilliType + TimestampMicroType + TimestampNanoType` instead?

sunchao replied:
Like mentioned above, if we agree to revamp the `DateType` for date, time, timestamp and interval, the name will be shorter.

nevi-me commented on rust/arrow/src/datatypes.rs:
```diff
@@ -156,6 +157,107 @@ make_type!(UInt64Type, u64, DataType::UInt64, 64, 0u64);
+    0i64
+);
+make_temporal_type!(
+    IntervalYearMonthType,
```
Interval isn't yet well defined in the format, so I was unsure of whether interval types should be backed by `i32` or `i64`.

paddyhoran commented:
I'm excited to get this merged, thanks @nevi-me.  No time right now but I'll definitely review over the weekend if the others don't beat me to it.

nevi-me commented:
Thanks @paddyhoran, I'd prefer if we keep the PR open long enough for Rust contributors to review (or look at) so we reach consensus on naming, to minimise refactors.

paddyhoran commented on rust/arrow/src/array.rs:
```diff
@@ -249,6 +263,116 @@ impl<T: ArrowNumericType> PrimitiveArray<T> {
+    /// If a data type cannot be converted to `NaiveDateTime`, a `None` is returned
+    /// TODO: extract constants into static variables
+    pub fn datetime(&self, i: usize) -> Option<NaiveDateTime> {
+        if !self.is_valid(i) {
```
I don't think you should check the validity of an array slot.  These methods are similar to `value` so you should allow the caller to check the validity of an array slot themselves.  i.e. the caller may know that there are no nulls and so would not want to check this on every call to this method.

In fact, I would consider changing these method names to `value_as_datetime`, `value_as_date` to make this connection.

nevi-me replied:
Thanks, I've updated the names and removed the validity check

paddyhoran commented on rust/arrow/src/array.rs:
```diff
@@ -784,6 +996,76 @@ mod tests {
+    #[test]
+    fn test_date64_array_from_vec_option() {
+        // Test building a primitive array with null values
+        // we use Int32 and Int64 as a backing array, so all Int32 and Int64 convensions work
```
typo in comments, should be `conversions` I think...  (repeated below)

nevi-me replied:
I meant `conventions`, thanks, I'll fix the typos

paddyhoran commented on rust/arrow/src/array.rs:
```diff
@@ -800,6 +1082,17 @@ mod tests {
 
+    #[test]
+    fn test_timestamp_fmt_debug() {
+        // TODO might be preferable to format these as datetime strings
```
Yes - I agree, datetime strings would be better.

nevi-me replied:
Found a way to do it with specialisation

nevi-me commented:
@andygrove please have a look at this when you get a chance. I'd like to include cast operations and CSV reader + writer for temporal arrays in the coming weeks, hopefully with 0.13

sunchao commented:
I'll take a look at this later today as well.

nevi-me commented:
@andygrove aside from the spec, are the array names fine? I felt they got long (`Time32NanosecondArray`)

wesm commented:
Was going to merge (have not looked closely) but will await @sunchao code review

sunchao commented on rust/arrow/src/datatypes.rs:
```diff
@@ -156,6 +157,107 @@ make_type!(UInt64Type, u64, DataType::UInt64, 64, 0u64);
 make_type!(Float32Type, f32, DataType::Float32, 32, 0.0f32);
 make_type!(Float64Type, f64, DataType::Float64, 64, 0.0f64);
 
+macro_rules! make_temporal_type {
```
Instead of this macro, can we reuse `make_type` - we'll need to move the 
```rust
impl ArrowNativeType for $native_ty {}
```
out though, but I think it's better than having a separate but almost the same macro.

sunchao commented on rust/arrow/src/array.rs:
```diff
@@ -249,8 +263,113 @@ impl<T: ArrowNumericType> PrimitiveArray<T> {
+    /// TODO: extract constants into static variables
+    pub fn value_as_datetime(&self, i: usize) -> Option<NaiveDateTime> {
+        let v = i64::from(self.value(i));
+        match &self.data_type() {
```
nit: do we need the `&` here? seems this:
```rust
match self.data_type() {
  DataType::Date32(_) => {
  ...
```
also work.

sunchao commented on rust/arrow/src/array.rs:
```diff
@@ -370,6 +524,94 @@ def_numeric_from_vec!(UInt32Type, u32, DataType::UInt32);
 def_numeric_from_vec!(Float64Type, f64, DataType::Float64);
+// TODO: add temporal arrays
+
+macro_rules! def_temporal_from_vec {
```
can we reuse `def_numeric_from_vec`? we just need to change `path` to `expr`. The macro name may also need to be changed.

nevi-me replied:
I didn't know that with `expr` and `path`, I would have reused the macros from the onset. Thanks!

nevi-me commented:
@sunchao I've updated

sunchao commented on rust/arrow/src/array.rs:
```diff
@@ -249,8 +263,113 @@ impl<T: ArrowNumericType> PrimitiveArray<T> {
+    ///
+    /// `Date32` and `Date64` return UTC midnight as they do not have time resolution
+    pub fn value_as_time(&self, i: usize) -> Option<NaiveTime> {
+        match &self.data_type() {
```
nit: you can also remove `&` here.

sunchao commented on rust/arrow/src/array.rs:
```diff
@@ -249,8 +263,113 @@ impl<T: ArrowNumericType> PrimitiveArray<T> {
+    ///
+    /// If a data type cannot be converted to `NaiveDateTime`, a `None` is returned.
+    /// A valid value is expected, thus the user should first check for validity.
+    /// TODO: extract constants into static variables
```
can we extract the constants in this PR?

nevi-me replied:
Looks like I was benefiting a lot from implicit conversions between signed and unsigned, adding constants. I thought I could quickly extract the constants, but I have to spend a bit more time and reason about my casting. I'll do that later or tomorrow

```rust
error[E0277]: cannot multiply `i32` to `u32`
   --> arrow\src\array.rs:293:43
    |
293 |                 (v % MILLISECONDS) as u32 * MICROSECONDS,
    |                                           ^ no implementation for `u32 * i32`
```

sunchao commented on rust/arrow/src/array.rs:
```diff
@@ -249,8 +263,113 @@ impl<T: ArrowNumericType> PrimitiveArray<T> {
+                Some(NaiveDateTime::from_timestamp(v * 86_400, 0))
+            }
+            DataType::Date64(_) => Some(NaiveDateTime::from_timestamp(
+                v / 1_000,
```
I think the `(v % 1_000) as u32 * 1_000_000` is not needed but either way is fine.

nevi-me replied:
I need the nanoseconds, otherwise I lose a bit of accuracy

nevi-me commented:
**Please don't merge yet, there's a few nits that I need to address**. Thanks

nevi-me commented:
@sunchao I've extracted constants, added a few inline comments and moved the temporal kernel to `compute/kernels/temporal.rs`. I'll fix the eventual merge conflict in the cast kernel if this gets merged before it.

sunchao commented:
Merged. Thanks @nevi-me !
